aiidalab/app.py

```diff
@@ -0,0 +1,647 @@
+
+        # Found local commits.
+        if self.found_local_commits():
+            return 'local commits found (risk to lose your work)'
+
+        # Found no branches.
+        if not self.available_versions:
+            return 'no branches found'
+
+        return ''
+
+    def git_update_available(self):
+        """Check whether there are updates available for the current branch in the remote repository."""
+
+        if self.current_version is None or not self._git_url or self.current_version.startswith(b'refs/tags/'):
+            # For later: if it is a tag check for the newer tags
+            to_return = False
+
+        # If it is a branch.
+        elif self.current_version.startswith(b'refs/remotes/'):
+
+            # Learn about local repository.
+            local_branch = re.sub(b'refs/remotes/(\w+)/', b'refs/heads/', self.current_version)  # pylint:disable=anomalous-backslash-in-string
+            local_head_at = self.repo[bytes(local_branch)]
+            remote_head_at = self.repo[bytes(self.current_version)]
+
+            # Learn about remote repository.
+            try:
+                on_server_head_at = self._git_remote_refs[local_branch]
+            except KeyError:
+                return False
+
+            # Is remote reference on the server is outdated?
+            if remote_head_at.id != on_server_head_at:
+
+                # I check if the remote commit is present in my remote commit history.
+                for cmmt in self.repo.get_walker(remote_head_at.id):
+                    if cmmt.commit.id == on_server_head_at:
+                        to_return = False
+                        break
+                else:
+                    to_return = True
+
+            elif local_head_at != remote_head_at:
+
+                # Go back in the current branch commit history and see if I can find the remote commit there.
+                for cmmt in self.repo.get_walker(local_head_at.id):
+
+                    # Found, so the remote branch is outdated - no update.
+                    if cmmt.commit.id == remote_head_at.id:
+                        to_return = False
+                        break
+
+                # Not found, so the remote branch has additional commit(s).
+                else:
+                    to_return = True
+            # else
+            #return False
+        # local branches can't have an update
+        #if self.current_version.startswith(b'refs/heads'):
+        #    return False
+        # something else
+        return to_return
+
+    @property
+    def install_button(self):
+        """Button to install the app."""
+        if not hasattr(self, '_install_button'):
+            self._install_button = ipw.Button(description="install")
+            self._install_button.on_click(self._install_app)
+            self._refresh_install_button()
+        return self._install_button
+
+    def _install_app(self, _):
+        """Installing the app."""
```

Comment:
I get the impression that this is class is handling too many concerns. It is representing the application itself, but also a stand-in, because in the former case we assume that it is something that *is installed*, and can be rendered etc., and in the second case it is something that *can be installed*. So the latter would be better handled by some `AiidaLabAppStoreEntry` class or so.

I propose to merge _as is_ and I am going to split these various concerns in my proposed revisions.

Reply:
I agree. Some time ago I've even made an issue (#28) about the Home App design. Happy to rediscuss that with you.